unit_scaling/_modules.py

```diff
@@ -198,35 +399,36 @@ class MHSA(nn.Module):
     Args:
         hidden_size (int): the hidden dimension size of the input.
         heads (int): the number of attention heads.
+        is_causal (bool): causal masking (for non-padded sequences).
         dropout_p (float, optional): the probability of the post-softmax dropout.
         {0}
+        {1}
     """
 
     def __init__(
         self,
         hidden_size: int,
         heads: int,
-        dropout_p: float,
-        constraint: Optional[str] = "gmean",
+        is_causal: bool,
+        dropout_p: float = 0.0,
+        mult: float = 1.0,
+        constraint: Optional[str] = "to_output_scale",
     ) -> None:
         super().__init__()
         self.heads = heads
         self.dropout_p = dropout_p
-        self.linear_qkv = Linear(
-            hidden_size, 3 * hidden_size, bias=False, constraint=constraint
-        )
-        self.linear_o = Linear(
-            hidden_size, hidden_size, bias=False, constraint=constraint
-        )
+        self.is_causal = is_causal
+        self.mult = mult
+        self.linear_qkv = Linear(hidden_size, 3 * hidden_size, constraint=constraint)
```

Comment:
Without a `linear` specialisation (like `linear_readout`), this will have an extra `sqrt(3)` factor in the scale.

Reply:
Ah, yes, I wondered about this; the cheeky thing I was thinking is that when using the new-default constraint "to_output_scale" it shouldn't matter, as the scale just depends on fan_in. But maybe this is a bit cheeky as it can be overridden.

Reply:
> the scale just depends on fan_in
Won't the output scale of this op be `3 * fan_in` rather than fan_in though? I see in the demo notebook that `attn_qkv.weight.grad.std = 0.62`, which is pretty close to `1/sqrt(3) = 0.58`.